extensions/grpc/src/test/java/com/hazelcast/jet/grpc/GrpcServiceTest.java

```diff
@@ -0,0 +1,316 @@
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.grpc;
+
+import com.hazelcast.jet.SimpleTestInClusterSupport;
+import com.hazelcast.jet.config.JetConfig;
+import com.hazelcast.jet.grpc.greeter.GreeterGrpc;
+import com.hazelcast.jet.grpc.greeter.GreeterOuterClass.HelloReply;
+import com.hazelcast.jet.grpc.greeter.GreeterOuterClass.HelloRequest;
+import com.hazelcast.jet.impl.util.ExceptionUtil;
+import com.hazelcast.jet.pipeline.BatchStage;
+import com.hazelcast.jet.pipeline.BatchStageWithKey;
+import com.hazelcast.jet.pipeline.Pipeline;
+import com.hazelcast.jet.pipeline.ServiceFactory;
+import com.hazelcast.jet.pipeline.Sinks;
+import com.hazelcast.jet.pipeline.test.AssertionSinks;
+import com.hazelcast.jet.pipeline.test.TestSources;
+import io.grpc.BindableService;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.Server;
+import io.grpc.ServerBuilder;
+import io.grpc.stub.StreamObserver;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.IntStream;
+
+import static com.hazelcast.jet.grpc.GrpcServices.bidirectionalStreamingService;
+import static com.hazelcast.jet.grpc.GrpcServices.unaryService;
+import static java.util.stream.Collectors.toList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class GrpcServiceTest extends SimpleTestInClusterSupport {
+
+    private static final int ITEM_COUNT = 10_000;
+    Server server;
+
+    @BeforeClass
+    public static void setup() {
+        initialize(2, new JetConfig());
+    }
+
+    @After
+    public void teardown() {
+        if (server != null) {
+            server.shutdown() ;
+        }
+    }
+
+    @Test
+    public void when_bidirectionalStreaming() throws IOException {
+        // Given
+        server = createServer(new GreeterServiceImpl());
+        final int port = server.getPort();
+
+        ServiceFactory<?, ? extends GrpcService<HelloRequest, HelloReply>> greeterService =
+                bidirectionalStreaming(port);
+
+        Pipeline p = Pipeline.create();
+        BatchStage<String> stage = p.readFrom(TestSources.items("one", "two", "three", "four"));
+
+        // When
+        BatchStage<String> mapped = stage.mapUsingServiceAsync(greeterService, (service, item) -> {
+            HelloRequest req = HelloRequest.newBuilder().setName(item).build();
+            return service.call(req).thenApply(HelloReply::getMessage);
+        });
```

Comment:
I think it is quite hard to work what happens when, what is input/output of what even on this toy example, the example with productInfo and brokerInfo is even harder to read.

IMHO the main issue is that the logical flow jumps back and forth in the source code
1. mapping from the pipeline to the request at line 84
1. service definition (including the actual operation on the stub) at line 76
1. mapping the service response at line 85

And the complicated functions returning consumers/functions etc.. don't help:
 ```
FunctionEx<? super ManagedChannel, ? extends FunctionEx<StreamObserver<R>, StreamObserver<T>>>
```

We should provide more syntactic sugar so it's easier to use for a regular developer. Something along the lines of:
```
        ServiceFactory<?, ? extends GrpcService<GreeterStub>> greeterService =
                bidirectionalStreamingService(
                        () -> ManagedChannelBuilder.forAddress("localhost", port).usePlaintext()
                );

        Pipeline p = Pipeline.create();
        BatchStage<String> stage = p.readFrom(TestSources.items("one", "two", "three", "four"));

        // When
        BatchStage<String> mapped = stage.mapUsingServiceAsync(greeterService,
                (item) -> HelloRequest.newBuilder().setName(item).build(),
                GreeterStub::sayHelloBidirectional,
                (response) -> response.getMessage() // Or just HelloReply::getMessage
        );
```